src/arena.c

```diff
@@ -605,6 +605,88 @@ arena_reset(tsd_t *tsd, arena_t *arena) {
 	pa_shard_reset(tsd_tsdn(tsd), &arena->pa_shard);
 }
 
+#define ARENA_DESTROY_MAX_DELAYED_MTX 32
+static void
+arena_prepare_base_deletion_sync_finish(tsd_t *tsd, malloc_mutex_t **mutexes,
+    unsigned n_mtx) {
+	for (unsigned i = 0; i < n_mtx; i++) {
+		malloc_mutex_lock(tsd_tsdn(tsd), mutexes[i]);
+		malloc_mutex_unlock(tsd_tsdn(tsd), mutexes[i]);
+	}
+}
+
+static void
+arena_prepare_base_deletion_sync(tsd_t *tsd, malloc_mutex_t *mtx,
+    malloc_mutex_t **delayed_mtx, unsigned *n_delayed) {
+	if (!malloc_mutex_trylock(tsd_tsdn(tsd), mtx)) {
+		/* No contention. */
+		malloc_mutex_unlock(tsd_tsdn(tsd), mtx);
+		return;
+	}
+	assert(*n_delayed < ARENA_DESTROY_MAX_DELAYED_MTX);
+	/* Add another to the batch. */
+	delayed_mtx[*n_delayed++] = mtx;
+
+	if (*n_delayed == ARENA_DESTROY_MAX_DELAYED_MTX) {
+		arena_prepare_base_deletion_sync_finish(tsd, delayed_mtx,
+		    *n_delayed);
+		n_delayed = 0;
+	}
+}
+
+static void
+arena_prepare_base_deletion(tsd_t *tsd, base_t *base_to_destroy) {
+	/*
+	 * In order to coalesce, emap_try_acquire_edata_neighbor will attempt to
+	 * check neighbor edata's state to determine eligibility.  This means
+	 * under certain conditions, the metadata from an arena can be accessed
+	 * w/o holding any locks from that arena.  In order to guarantee safe
+	 * memory access, the metadata and the underlying base allocator needs
+	 * to be kept alive, until all pending accesses are done.
+	 *
+	 * 1) with opt_retain, the arena boundary implies the is_head state
+	 * (tracked in the rtree leaf), and the coalesce flow will stop at the
+	 * head state branch.  Therefore no cross arena metadata access
+	 * possible.
+	 *
+	 * 2) w/o opt_retain, the arena id needs to be read from the edata_t,
+	 * meaning read only cross-arena metadata access is possible.  The
+	 * coalesce attempt will stop at the arena_id mismatch, and is always
+	 * under one of the ecache locks.  To allow safe passthrough of such
+	 * metadata accesses, the loop below will iterate through all manual
+	 * arenas' ecache locks.  As all the metadata from this base allocator
+	 * have been unlinked from the rtree, after going through all the
+	 * relevant ecache locks, it's safe to say that a) pending accesses are
+	 * all finished, and b) no new access will be generated.
+	 */
+	if (opt_retain) {
+		return;
+	}
+	unsigned destroy_ind = base_ind_get(base_to_destroy);
+	assert(destroy_ind >= manual_arena_base);
+
+	tsdn_t *tsdn = tsd_tsdn(tsd);
+	malloc_mutex_t *delayed_mtx[ARENA_DESTROY_MAX_DELAYED_MTX];
+	unsigned n_delayed = 0;
+	for (unsigned i = 0; i < narenas_total_get(); i++) {
```

Comment:
@davidtgoldblatt what do you think of this approach? Slightly tweaked based on what we discussed: instead of tracking the is_manual_arena / is_destroyable state in rtree leaf, here we ask the destroy operation to go through all arenas' locks, versus only the manual arenas'. Since this can make the # of locks involved a lot higher, I added a simple trylock + batching mechanism to compensate.

(only the commit adding this part is new (and rebased to the earliest), other commits not updated yet.

Reply:
I like just acquiring all the locks.

The trylock/batching thing feels like premature optimization to me; do we care that much about arena destroy latency / do we even have a benchmark implying this would help[1]? Not really opposed per se, but it feels like an edge-casey thing to take on the code complexity for.

[1] You could imagine a world where if an ecache lock is hot, we're always gonna have to do a contended spin to acquire it. So the initial trylock pointlessly slows down other threads without the benefit. I'm not saying we live in this world, but without a use case it's hard to even be sure if batching helps/hurts.

Reply:
Yeah it's not a straight win. My thoughts were: in the case where it ends up hurting, it should be just another small dent in an already terrible situation, and won't really gets blamed for the whole issue; but in the case where it helps (e.g. a busy workload resulting semi-hot ecache locks), the out of order locking could make the overhead / latency tolerable.

Though the truth is I don't really have a preference; I admit the initial motivation was like, oh I made the this path slower, let's see if there's any remedy for that. The final call is your @davidtgoldblatt :)

Reply:
Eh fair enough. I feel like status quo wins in a situation where no one has a strong preference, so I'm fine leaving it the way it is now.

Reply:
Haha okay. I'll update the PR and let you know when it's ready.